sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/AvroUtils.java

```diff
@@ -20,87 +20,210 @@
 
   /**
    * Converts AVRO schema to Beam row schema.
    *
    * @param schema schema of type RECORD
    */
-  public static Schema toSchema(@Nonnull org.apache.avro.Schema schema) {
+  public static Schema toBeamSchema(org.apache.avro.Schema schema) {
     Schema.Builder builder = Schema.builder();
 
     for (org.apache.avro.Schema.Field field : schema.getFields()) {
-      org.apache.avro.Schema unwrapped = unwrapNullableSchema(field.schema());
+      TypeWithNullability nullableType = new TypeWithNullability(field.schema());
+      builder.addField(field.name(), toFieldType(nullableType));
+    }
+
+    return builder.build();
+  }
+
+  /** Converts a Beam Schema into an AVRO schema. */
+  public static org.apache.avro.Schema toAvroSchema(Schema beamSchema) {
+    List<org.apache.avro.Schema.Field> fields = Lists.newArrayList();
+    for (Schema.Field field : beamSchema.getFields()) {
+      org.apache.avro.Schema fieldSchema = getFieldSchema(field.getType());
+      org.apache.avro.Schema.Field recordField =
+          new org.apache.avro.Schema.Field(
+              field.getName(), fieldSchema, field.getDescription(), null);
+      fields.add(recordField);
+    }
+    org.apache.avro.Schema avroSchema = org.apache.avro.Schema.createRecord(fields);
+    return avroSchema;
+  }
+
+  /**
+   * Strict conversion from AVRO to Beam, strict because it doesn't do widening or narrowing during
+   * conversion. If Schema is not provided, one is inferred from the AVRO schema.
+   */
+  public static Row toBeamRowStrict(GenericRecord record, @Nullable Schema schema) {
+    if (schema == null) {
+      schema = toBeamSchema(record.getSchema());
+    }
+
+    Row.Builder builder = Row.withSchema(schema);
+    org.apache.avro.Schema avroSchema = record.getSchema();
+
+    for (Schema.Field field : schema.getFields()) {
+      Object value = record.get(field.getName());
+      org.apache.avro.Schema fieldAvroSchema = avroSchema.getField(field.getName()).schema();
 
-      if (!unwrapped.equals(field.schema())) {
-        builder.addNullableField(field.name(), toFieldType(unwrapped));
+      if (value == null) {
+        builder.addValue(null);
       } else {
-        builder.addField(field.name(), toFieldType(unwrapped));
+        builder.addValue(convertAvroFieldStrict(value, fieldAvroSchema, field.getType()));
       }
     }
 
     return builder.build();
   }
 
+  /**
+   * Convert from a Beam Row to an AVRO GenericRecord. If a Schema is not provided, one is inferred
+   * from the Beam schema on the orw.
+   */
+  public static GenericRecord toGenericRecord(
+      Row row, @Nullable org.apache.avro.Schema avroSchema) {
+    Schema beamSchema = row.getSchema();
+    // Use the provided AVRO schema if present, otherwise infer an AVRO schema from the row
+    // schema.
+    if (avroSchema != null && avroSchema.getFields().size() != beamSchema.getFieldCount()) {
+      throw new IllegalArgumentException("AVRO schema doesn't match row schema");
```

Comment:
nit: would be nice to have include schemas into error message as in `wrapIfNullable`